### ISSUE TITLE
[Feature Request] MaskableRecurrentPPO

### ISSUE THREAD

**Issue by CppMaster** [Stable-Baselines-Team/stable-baselines3-contrib#101]
**Motivation**
MaskablePPO is great for large discrete action space that has many invalid actions at each step, while RecurrentPPO is useful for the agent to has a memory of previous observations and actions taken, which improves it's decision making. Right now, we have to choose between those 2 algorithms and cannot have features of both of them, which would greatly improve agents training when both action masking and sequence processing is helpful.

**Feature**
MaskableRecurrentPPO - An algorithm that is a combination of MaskablePPO and RecurrentPPO. Or action masking integration to PPO and RecurrentPPO.

**Comment by araffin** [comment]
Duplicate of https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/issues/76

@dylanprins I would happy to share the link in the doc if you could open source your implementation ;)

**Comment by rextea** [comment]
@dylanprins +1
If anyone has a solution for it, it can be really great :)

**Comment by rllyryan** [comment]
I would love to try this implementation if it is available! MaskablePPO is the only algorithm that makes use of the masking technique :+1:

**Comment by philippkiesling** [comment]
@rllyryan Are you currently working on this? 

If not, I can do this, since I haven't found any implementations of this so far :) 

@araffin Should I do a PR for this or would you prefer a separate repository for this?

**Comment by araffin** [comment]
> hould I do a PR for this or would you prefer a separate repository for this?

I would prefer a separate repo but we would put a link to it in our doc.

**Comment by rllyryan** [comment]
Hey @philippkiesling, I apologise for the late reply, I was focusing on tuning my custom environment (for work) for a while now, did not have the flexibility to carry out the merging. That being said, I just started out in this field, and am unlikely to have the proficiency to make this modification myself. Thank you for taking this up!

I will try to take a look at your first implementation, and understand how this merger is going to work (merge between RPPO and MaskablePPO). Currently, my environment is having Box inputs for MaskablePPO.

**Comment by philippkiesling** [comment]
Hi, I needed this for a paper myself, so no additional work for me :) 
I tried box inputs, and it should work, but please let me know if there are any problems

**Comment by rllyryan** [comment]
Sure! I will try it out in the next few days :) Will let you know if it ran to completion!

**Comment by wdlctc** [comment]
Update newest recurrentmaskable PPO on https://github.com/wdlctc/recurrent_maskable based on sb3-contrib version 1.8.0 since the previous version can't work on current version of sb3-contrib. Feel free to take it.

**Comment by araffin** [comment]
> Update newest recurrentmaskable PPO on https://github.com/wdlctc/recurrent_maskable based on sb3-contrib version 1.8.0 

Thanks for the update, do you mean it now works with current SB3 master version?

**Comment by wdlctc** [comment]
> > Update newest recurrentmaskable PPO on https://github.com/wdlctc/recurrent_maskable based on sb3-contrib version 1.8.0
> 
> Thanks for the update, do you mean it now works with current SB3 master version?

Yes, up to now it works well on stable_baselines3 version 1.8.0 on my custom environment and public test environment.

**Comment by tty666** [comment]
both are from last year without update ... someone could take over to make it available for latest versions ?

**Comment by Maxxxel** [comment]
+1 would love to see an update for 2.3.* ❤️

**Comment by akbaig** [comment]
@araffin Hi, I'm seeing a lot of people requiring this feature since 2022. Although it's understandable to have a clean repo, we can already notice the dire need i.e. two independent repos were created but never maintained to keep up with the newer versions. I'm sure there are other people who searched for this feature and never found it. Therefore, I think it's time to take the leap of faith. Thinking of implementing it for the newer version. Thoughts?

**Comment by dahovey** [comment]
@akbaig @araffin +1 for a Maskable RecurrentPPO implementation here.

How complex is it to add masking? I am not a math wiz.

**Comment by kirianguiller** [comment]
Any news on this feature request ? Thank you for your time
